clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```diff
@@ -41,26 +44,45 @@
     private final SubscriptionState subscriptions;
     private Map<TopicPartition, List<ConsumerRecord<K, V>>> records;
     private boolean closed;
+    private final Map<TopicPartition, Long> beginningOffsets;
+    private final Map<TopicPartition, Long> endOffsets;
+
+    private AtomicReference<CountDownLatch> pollLatch;
+    private KafkaException exception;
+
+    private AtomicBoolean wakeup;
 
     public MockConsumer(OffsetResetStrategy offsetResetStrategy) {
         this.subscriptions = new SubscriptionState(offsetResetStrategy);
         this.partitions = new HashMap<String, List<PartitionInfo>>();
         this.records = new HashMap<TopicPartition, List<ConsumerRecord<K, V>>>();
         this.closed = false;
+        this.beginningOffsets = new HashMap<>();
+        this.endOffsets = new HashMap<>();
+        this.pollLatch = new AtomicReference<>();
+        this.exception = null;
+        this.wakeup = new AtomicBoolean(false);
     }
     
     @Override
-    public synchronized Set<TopicPartition> assignment() {
+    public Set<TopicPartition> assignment() {
         return this.subscriptions.assignedPartitions();
     }
 
+    /** Simulate a rebalance event. */
+    public void rebalance(Collection<TopicPartition> newAssignment) {
+        // TODO: Rebalance callbacks
```

Comment:
The offset storage doesn't require this since it uses simple consumer mode -- each worker is just reading the entire topic. If nobody adds this to MockConsumer in the meantime, I'll have to add it in KAFKA-2371 (distributed herder). Just trying to keep these patches smaller and incremental.

To be honest, I wasn't sure how useful the MockConsumer was going to be and whether we'd really use it externally. I'll be interested to see whether/how KStreams uses it. In this patch, I originally started with Easy/PowerMock mocks, but they were kind of hard to make work well given that I had a multithreaded usage here.